datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt

```diff
@@ -154,10 +232,76 @@ private val Questionnaire.itemContextNameToExpressionMap: Map<String, String>
       .toMap()
   }
 
+/**
+ * Extracts a list containing the resource name followed by field names leading to the destination
+ * field defined in the [definition] field, or `null` if the [definition] field is empty or invalid.
+ *
+ * For example, if the [definition] field is
+ * "http://hl7.org/fhir/StructureDefinition/Patient#Patient.name", `listOf("Patient", "name")` will
+ * be returned.
+ */
+private val Questionnaire.QuestionnaireItemComponent.targetResourceAndElement: List<String>?
+  get() {
+    val snapshotPath = definition.substringAfter('#', "")
+    if (!"[a-zA-Z]+(\\.[a-zA-Z]+)+".toRegex().matches(snapshotPath)) return null
+
+    return snapshotPath.split(".")
+  }
+
+/**
+ * Retrieves details about the target field defined in
+ * [org.hl7.fhir.r4.model.Questionnaire.QuestionnaireItemComponent.definition] for easier searching
+ * of the setter methods and converting the answer to the expected parameter type.
+ */
+private val Questionnaire.QuestionnaireItemComponent.inferPropertyResourceClass: Field?
+  get() {
+    val modelAndField = targetResourceAndElement ?: return null
+    var resourceClass: Class<*> = Class.forName("org.hl7.fhir.r4.model.${modelAndField.get(0)}")
+    var resourceField: Field? = null
+
+    modelAndField.forEachIndexed loop@{ index, fieldName ->
+      if (index == 0) return@loop
+
+      resourceField = resourceClass.getFieldOrNull(fieldName) ?: return null
+      resourceClass = resourceField!!.mainType
+    }
+
+    return resourceField
```

Comment:
try something like this:

```suggestion
    val path = definitionPath ?: return null
    if (path.size < 2) return null
    var resourceClass: Class<*> = Class.forName("org.hl7.fhir.r4.model.${path[0]}")
    val field: Field = resourceClass.getField(path[1])
    return path.drop(2).fold(field) { field: Field, nestedFieldName: String ->
      field.mainType.getField(nestedFieldName)
    }
```

i think the `fold` function in kotlin is perfect for this.

Reply:
Tried to re-write it this way. The results

```
	val path = definitionPath ?: return null
    if (path.size < 2) return null
    val resourceClass: Class<*> = Class.forName("org.hl7.fhir.r4.model.${path[0]}")
    val field: Field = resourceClass.getFieldOrNull(path[1]) ?: return null

    return path.drop(2).fold(field) { field: Field?, nestedFieldName: String ->
      field?.nonParameterizedType?.getFieldOrNull(nestedFieldName)
    }
```